filegenerator/file_generator.go

```diff
@@ -0,0 +1,368 @@
+// permissions and limitations under the License.
+//
+// Author: William Haack (will@cockroachlabs.com)
+
+// The purpose of filegenerator is to provide the tools for the automatic
+// deployment of JSON files that are used on the benchviz web server
+// as the data to be displayed. RenderHistoricalBenchmarkResults
+// creates a BenchPackages struct that has all the data needed to
+// generate most of the JSON files. The various other methods use
+// the BenchPackages created by RenderHistoricalBenchmarkResults
+// in order to create the JSON files.
+
+package filegenerator
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"log"
+	"math"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+)
+
+const (
+	deployRootEnv    string = "BENCHDEPLOY"
+	benchSamplesEnv  string = "BENCHSAMPLES"
+	awsBucketNameEnv string = "AWSBUCKETNAME"
+	cockroachRootEnv string = "MYROOT"
+)
+
+// BenchPackages is a mapping of package names to test names, which in turn map
+// to their historical data.
+type BenchPackages map[string]BenchTestMap
+
+// BenchTestMap is a mapping of test names to dates, which in turn map to the
+// results of the benchmark tests for that given date.
+type BenchTestMap map[string]BenchResults
+
+// BenchResults map dates to the benchmark test data for that date.
+type BenchResults map[string]BenchStats
+
+// BenchStats is a struct containing the various metrics obtained from running
+// a benchmark test with the -benchmem parameter. The variable names are kept
+// short to keep the JSON files small.
+type BenchStats struct {
+	N int     // ns/op
+	A int     // alllocs/op
+	B int     // B/op
+	M float64 // MB/s
+}
+
+// GeometricMeanData is a struct that contains the means of the benchmark
+// stats for all the tests in a given package.
+type GeometricMeanData struct {
+	NMean float64 // Mean of the ns/op result from all tests in a package.
+	AMean float64 // Mean of the allocs/op result from all tests in a package.
+	BMean float64 // Mean of the B/op result from all tests in a package.
+	MMean float64 // Mean of the MB/s result from all tests in a package.
+	Date  string  // The date that the mean of all tests was collected.
+}
+
+func isValidTestLine(line string) bool {
+	matched, err := regexp.MatchString("FAIL", line)
+	if err != nil {
+		log.Fatal(err)
+	}
+	// This matches a line from Stdout where a benchmark test passes.
+	return len(line) >= 9 && line[0:9] == "Benchmark" && !matched
+}
+
+func parseTestResultLine(line string) (string, BenchStats) {
+	testName := strings.Trim(strings.Split(line, "\t")[0], " ")
+	nsopr, _ := regexp.Compile("[0-9]+ ns/op")
+	allocpr, _ := regexp.Compile("[0-9]+ allocs/op")
+	bopr, _ := regexp.Compile("[0-9]+ B/op")
+	mbpsr, _ := regexp.Compile("[0-9]+\\.?[0-9]+ MB/s")
+	nsops := nsopr.FindString(line)
+	allocps := allocpr.FindString(line)
+	bops := bopr.FindString(line)
+	mbpss := mbpsr.FindString(line)
+	var nsopi, allocpi, bopi int
+	var mbpsf float64
+	if nsops != "" {
```

Comment:
It'd be cleaner if you structured this code like this:

```
if nsops := nsopr.FindString(line); nsops != "" {
  ...
} else if allocps := allocpr.FindString(line); allocps != "" {
  ...
}
```

This is more compact and reduces the scope of `allocps`, etc. to what's strictly necessary.